WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -69,6 +62,15 @@ def index
       @competitions = @competitions.select { |competition| competition.contains?(params[:search]) }
     end
 
+    unless params[:event_ids].empty?
+      @competitions = @competitions.select do |competition|
+        (competition.eventSpecs.split & params[:event_ids]).any?
+      end.sort_by do |competition|
+        -(competition.eventSpecs.split & params[:event_ids]).count
```

Comment:
I've done a benchmark in rails console.
```ruby
ATTEMPS = 1000
event_ids = ["333", "444", "555", "666"]
competitions = Competition.where(showAtAll: true).order(:year, :month, :day).reverse_order
#########################################################
# My solution

overall_first = 0
ATTEMPS.times do
  overall_first += Benchmark.realtime do
    selected_competitions = competitions.select do |competition|
      (competition.eventSpecs.split & event_ids).any?
    end.sort_by! do |competition|
      -(competition.eventSpecs.split & event_ids).count
    end
  end
end

# Jeremy's solution

overall_second = 0
ATTEMPS.times do
  required_events = event_ids.map { |event_id| Event.find(event_id) }

  overall_second += Benchmark.realtime do
    competitions_and_matching_event_count = competitions.map { |competition| [competition, (competition.events & required_events).count] }
    competitions_and_matching_event_count.select { |competition_and_matching_event_count| competition_and_matching_event_count[1] > 0 }
    competitions_and_matching_event_count.sort_by! { |competition_and_matching_event_count| -competition_and_matching_event_count[1] }
    selected_competitions = competitions_and_matching_event_count.map! { |competition_and_matching_event_count| competition_and_matching_event_count[0] }
  end
end

# Jeremy's solution with event ids only

overall_third = 0
ATTEMPS.times do
  overall_third += Benchmark.realtime do
    competitions_and_matching_event_count = competitions.map { |competition| [competition, (competition.eventSpecs.split & event_ids).count] }
    competitions_and_matching_event_count.select { |competition_and_matching_event_count| competition_and_matching_event_count[1] > 0 }
    competitions_and_matching_event_count.sort_by! { |competition_and_matching_event_count| -competition_and_matching_event_count[1] }
    selected_competitions = competitions_and_matching_event_count.map! { |competition_and_matching_event_count| competition_and_matching_event_count[0] }
  end
end

# Summary

puts "First solution average of #{ATTEMPS}: #{overall_first / ATTEMPS}"
puts "Second solution average of #{ATTEMPS}: #{overall_second / ATTEMPS}"
puts "Third solution average of #{ATTEMPS}: #{overall_third / ATTEMPS}"
```
Results:
```
First solution average of 1000: 0.015748224090930307
Second solution average of 1000: 0.02872117611132853
Third solution average of 1000: 0.006999585387093248
```
So as average of 1000 indicated, when we join your suggestion with using event ids only, it becomes super fast.
Finally it would look like this (all tests passes yay):
```ruby
    unless params[:event_ids].empty?
      competitions_and_matching_event_count = @competitions.map do |competition|
        [competition, (competition.eventSpecs.split & params[:event_ids]).count]
      end
      competitions_and_matching_event_count.select! do |competition_and_matching_event_count|
        competition_and_matching_event_count[1] > 0
      end
      competitions_and_matching_event_count.sort_by! do |competition_and_matching_event_count|
        -competition_and_matching_event_count[1]
      end
      @competitions = competitions_and_matching_event_count.map do |competition_and_matching_event_count|
        competition_and_matching_event_count[0]
      end
    end
```

Or even (not sure which is cleaner):
```ruby
    unless params[:event_ids].empty?
      @competitions = @competitions.map do |competition|
        [competition, (competition.eventSpecs.split & params[:event_ids]).count]
      end.select do |competition_and_matching_event_count|
        competition_and_matching_event_count[1] > 0
      end.sort_by! do |competition_and_matching_event_count|
        -competition_and_matching_event_count[1]
      end.map! do |competition_and_matching_event_count|
        competition_and_matching_event_count[0]
      end
    end
```

Thoughts?